### PR TITLE
Train speech language ID classification head

### PR DESCRIPTION
**Pull request by am831**
Implements a new training script for training a classification head. The new nn.module in model.py is separate from other components of M4T and takes speech encoder output and maps it into language probabilities with softmax. All layers of M4T are frozen to prevent training of those components. A new dataset.py script is implemented for downloading common voice from hugging face.

TODO: 
- Write a new dataloader that discards transcript metadata and only prepares audio features and language labels
- Obtain results from training

I am collaborating with @zrthxn on this project

### CODE REVIEW

File: src/seamless_communication/cli/m4t/classification_head/model.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+    def forward(self, x):
+        if self.layers is None:
+            self.input_dim = x.size(-1)
+            self.hidden_dim = self.input_dim
+            self.layers = nn.Sequential(
+                nn.Linear(self.input_dim, self.hidden_dim),
+                nn.ReLU(),
+                nn.Linear(self.hidden_dim, self.num_languages)
+            )
```
Comment (zrthxn) [inline src/seamless_communication/cli/m4t/classification_head/model.py:20]: I just noticed this, you're actually creating new linear layers on every forward pass. This will cause new weights to be initialized every time you run the head which means it can never be trained. This should actually be done in the constructor.
I've fixed this in my PR to this branch.

Comment (am831) [reply src/seamless_communication/cli/m4t/classification_head/model.py:20]: Thanks for catching this

File: src/seamless_communication/cli/m4t/classification_head/train.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+            params: ClassificationHeadTrainParams,
+            label_weights: torch.Tensor = None):
+    
+    logger.info("Start Training Language Head")
+    dataloader = dataloader.get_dataloader()
+    frozen_model.train()
+    
+    grad_scaler = torch.cuda.amp.GradScaler()
+    optimizer = AdamW(
+        params=frozen_model.parameters(),
```
Comment (zrthxn) [inline src/seamless_communication/cli/m4t/classification_head/train.py:174]: We want to optimize the head not the frozen model

File: src/seamless_communication/cli/m4t/classification_head/train.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+    parser.add_argument(
+        "--num_layers", 
+        type=int, 
+        default=2, 
+        help="The number of layers in the classification head"
+    )
+    return parser
+
+
+def plot_losslog(losslog: list, save_to: Path = None, yscale = 'log'):
```
Comment (mavlyutovr) [inline src/seamless_communication/cli/m4t/classification_head/train.py:141]: use tensorboard

File: src/seamless_communication/cli/m4t/classification_head/train.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+def main() -> None:
+    args = init_parser().parse_args()
+    device = torch.device(args.device)
+    float_dtype = torch.float16 if torch.device(args.device).type != "cpu" else torch.bfloat16
+    
+    text_tokenizer = load_unity_text_tokenizer(args.model_name)
+    unit_tokenizer = load_unity_unit_tokenizer(args.model_name)
+    
+    # Freeze everything in the model, only train classification head
+    model = load_unity_model(args.model_name, device=torch.device("cpu"), dtype=torch.float32)
```
Comment (mavlyutovr) [inline src/seamless_communication/cli/m4t/classification_head/train.py:239]: you still need to switch the model in train mode

File: src/seamless_communication/cli/m4t/classification_head/train.py [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+                assert seqs.src_tokens is not None
+                optimizer.zero_grad()
+                seqs.src_tokens = seqs.src_tokens.to(params.device)
+                labels = labels.to(params.device)
+                
+                with torch.autocast(device_type=params.device.type, dtype=params.float_dtype):
+                    mask = PaddingMask(seqs.src_lengths, seqs.src_tokens.size(1)).to(params.device)
+                    vector, _ = frozen_model.encode(seqs.src_tokens, padding_mask=mask)
+                
+                probs = head(vector)
```
Comment (mavlyutovr) [inline src/seamless_communication/cli/m4t/classification_head/train.py:195]: why not under autocast?

File: src/seamless_communication/cli/m4t/classification_head/model.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+            [ nn.Linear(embed_dim, embed_dim) for _ in range(n_layers) ] + \
+            [ nn.Linear(embed_dim, n_classes) ])
+
+    def forward(self, x):
+        # (Batch, Seq, Embed)
+        x, _ = self.attn(x, x, x)
+        x = x[:, 0]
+        for layer in self.layers:
+            x = nn.functional.relu(layer(x))
+        return nn.functional.softmax(x).float()
```
Comment (mavlyutovr) [inline src/seamless_communication/cli/m4t/classification_head/model.py:20]: you don't need softmax here, as cross_entropy expects unnormalized logits

File: src/seamless_communication/cli/m4t/classification_head/train.py [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+                
+                loss = torch.nn.functional.cross_entropy(labels, probs)
```
Comment (mavlyutovr) [inline src/seamless_communication/cli/m4t/classification_head/train.py:197]: it has to be the opposite order of arguments: proba, labels (see documentation)

Comment (mavlyutovr) [reply src/seamless_communication/cli/m4t/classification_head/train.py:197]: also add label smoothing (0.1-0.2)

File: src/seamless_communication/cli/m4t/classification_head/model.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
+from torch import nn
+
+class ClassificationHead(nn.Module):
+    def __init__(self, embed_dim, n_layers, n_classes):
+        super(ClassificationHead, self).__init__()
+        self.num_languages = n_classes
+        self.num_layers = n_layers
+        
+        self.attn = nn.MultiheadAttention(embed_dim, num_heads=1)
```
[the remark this answers is not in the document]
Comment (zrthxn) [reply src/seamless_communication/cli/m4t/classification_head/model.py]: Added an parameter for this

File: src/seamless_communication/cli/m4t/classification_head/train.py (hunk 5 above)
Comment (zrthxn) [reply src/seamless_communication/cli/m4t/classification_head/train.py]: Changed to under autocast

File: src/seamless_communication/cli/m4t/classification_head/model.py (hunk 6 above)
Comment (zrthxn) [reply src/seamless_communication/cli/m4t/classification_head/model.py]: Done

File: src/seamless_communication/cli/m4t/classification_head/train.py (hunk 7 above)
Comment (zrthxn) [reply src/seamless_communication/cli/m4t/classification_head/train.py]: Done both
